### PR TITLE
Improve Per-Monitor DPI awareness in more controls

### PR DESCRIPTION
**Pull request by MaartenBent**
Another batch of PM-DPI fixes, mostly cherry-picked from #334, tested with the samples.

The dataview sample is mostly ok, except for the text that uses `wxMarkupParser`. This will be fixed in the next PR with the `wxDC` changes.

### CODE REVIEW

File: include/wx/msw/private/dpiaware.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        wx/msw/private/dpiaware.h
+// Purpose:     AutoSystemDpiAware class
+// Author:      Maarten Bent
+// Created:     10/6/2016
```
Comment (vadz) [inline include/wx/msw/private/dpiaware.h]: Extremely minor, but I'd prefer to standardize on ISO 8601 format (YYYY-MM-DD) everywhere. In this particular case, I don't even know which date does this refer to, really.

File: include/wx/msw/private/dpiaware.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+#ifndef _WX_MSW_DPI_AWARE_H_
+#define _WX_MSW_DPI_AWARE_H_
+
+#ifndef WX_PRECOMP
+    #include "wx/msw/missing.h"
+#endif
+
+#include "wx/dynlib.h"
+
```
Comment (vadz) [inline include/wx/msw/private/dpiaware.h]: Maybe put this class inside `namespace wxMSWImpl` too?

File: samples/toolbar/toolbar.cpp [hunk 3]
```diff
@@ -307,7 +307,8 @@ bool MyApp::OnInit()
     // Create the main frame window
     MyFrame* frame = new MyFrame((wxFrame *) NULL, wxID_ANY,
                                  "wxToolBar Sample",
-                                  wxPoint(100, 100), wxSize(650, 350));
+                                  wxPoint(100, 100), wxDefaultSize);
+    frame->SetSize(frame->FromDIP(wxSize(650, 350)));
```
Comment (vadz) [inline samples/toolbar/toolbar.cpp]: Just a note (i.e. this is not related to this PR per se):

Seeing it like this, it does look ugly. And I'm not even sure if it's identical to the old code, we'd need to use `SetInitialSize()` to really have the same effect.

I think we may need to add a new `wxFrame` ctor allowing to specify the size in DIPs.

File: src/generic/animateg.cpp [hunk 4]
```diff
@@ -327,7 +327,7 @@ wxSize wxAnimationCtrl::DoGetBestSize() const
     if (m_animation.IsOk() && !this->HasFlag(wxAC_NO_AUTORESIZE))
         return m_animation.GetSize();
 
-    return wxSize(100, 100);
+    return FromDIP(wxSize(100, 100));
```
Comment (vadz) [inline src/generic/animateg.cpp]: I also wonder if we shouldn't provide `FromDIP(x, y)` overload. It may make things a bit confusing, but writing `FromDIP(100, 100)` would be more palatable.

Or maybe add `SizeFromDIP()` or even just `SizeInDIP()`?

(this is just another note/thought without direct effect on this PR).

File: src/generic/datavgen.cpp [hunk 5]
```diff
@@ -5251,6 +5254,28 @@ void wxDataViewCtrl::OnSize( wxSizeEvent &WXUNUSED(event) )
...
+    {
+        m_clientArea->ClearRowHeightCache();
+        m_clientArea->SetRowHeight(m_clientArea->GetDefaultRowHeight());
+    }
+
+    for ( unsigned i = 0; i < m_cols.size(); ++i )
+    {
+        int minWidth = m_cols[i]->GetMinWidth();
+        if ( minWidth > 0 )
+            minWidth = minWidth * event.GetNewDPI().x / event.GetOldDPI().x;
```
Comment (vadz) [inline src/generic/datavgen.cpp]: I'm on a roll with the new API ideas: it looks like having `wxDPIChangedEvent::Rescale()` could be useful too. And it probably should use `wxMulDivInt32()` internally and not just integer division.

File: src/msw/listctrl.cpp [hunk 6]
```diff
@@ -421,6 +425,46 @@ WXDWORD wxListCtrl::MSWGetStyle(long style, WXDWORD *exstyle) const
...
+        // will detect the font change
+        SetHeaderAttr(wxItemAttr());
+        SetHeaderAttr(item);
+    }
+}
+
+void wxListCtrl::OnDPIChanged(wxDPIChangedEvent &event)
+{
+    const int numCols = GetColumnCount();
+    for ( int i = 0; i < numCols; ++i ) {
```
Comment (vadz) [inline src/msw/listctrl.cpp]:
```suggestion
    for ( int i = 0; i < numCols; ++i )
    {
```

File: src/msw/listctrl.cpp [hunk 7]
```diff
@@ -953,6 +1002,12 @@ bool wxListCtrl::SetItem(wxListItem& info)
                 data->attr->AssignFrom(attrNew);
             else
                 data->attr = new wxItemAttr(attrNew);
+
+            if ( data->attr->HasFont() ) {
```
Comment (vadz) [inline src/msw/listctrl.cpp]:
```suggestion
            if ( data->attr->HasFont() )
            {
```

File: src/msw/toolbar.cpp [hunk 8]
```diff
@@ -1895,6 +1897,52 @@ void wxToolBar::OnEraseBackground(wxEraseEvent& event)
...
+            continue;
+
+        if ( wxControl* const control = tool->GetControl() )
+        {
+            const wxSize oldSize = control->GetSize();
+            wxSize newSize = oldSize * scaleFactor;
+
+            // choice based controls seem to automatically adjust their height
+            // when the font size increases, keep this size
+            if ( dynamic_cast<wxChoiceBase*>(control) && scaleFactor > 1 )
```
Comment (vadz) [inline src/msw/toolbar.cpp]: We should use `wxDynamicCast()` here, some people disable RTTI when building wxMSW.

But beyond this, it would be nice to avoid the cast by adding some virtual `MSWAdjustSizeOnDPIChange()` that would just call `SetSize()` by default but could be overridden in `wxChoice` to only adjust the width and not the height.

File: src/msw/toolbar.cpp [hunk 9]
```diff
@@ -1895,6 +1897,52 @@ void wxToolBar::OnEraseBackground(wxEraseEvent& event)
...
+        }
+
+        if ( wxStaticText* const staticText = tool->GetStaticText() )
+        {
+            // Use the best size for the label
+            staticText->SetSize(staticText->GetBestSize());
+        }
+    }
+
+    // Use CallAfter because creating the toolbar directly does sometimes not
```
Comment (vadz) [inline src/msw/toolbar.cpp]:
```suggestion
    // Use CallAfter because creating the toolbar directly sometimes doesn't
```

File: include/wx/msw/private/dpiaware.h [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+// ----------------------------------------------------------------------------
+// Temporarily change the DPI Awareness context to System
+// ----------------------------------------------------------------------------
+
+class AutoSystemDpiAware
+{
+    #define WXDPI_AWARENESS_CONTEXT_UNAWARE      ((WXDPI_AWARENESS_CONTEXT)-1)
+    #define WXDPI_AWARENESS_CONTEXT_SYSTEM_AWARE ((WXDPI_AWARENESS_CONTEXT)-2)
+    typedef WXDPI_AWARENESS_CONTEXT
```
Comment (vadz) [inline include/wx/msw/private/dpiaware.h]: I'm a bit confused: where is `WXDPI_AWARENESS_CONTEXT` defined? I can't find it anywhere...

**Review by vadz — COMMENTED** [review]
Thanks a lot, no real comments except the one about virtualizing the size adjustment after DPI change -- I think it might make sense to do it now, but please let me know if you disagree.

Thanks again!

P.S. `wx/dataview.h` doesn't seem to contain any significant changes, but I can just revert it myself when merging.

File: include/wx/msw/private/dpiaware.h (hunk 10 above)
Comment (MaartenBent) [reply include/wx/msw/private/dpiaware.h]: https://github.com/wxWidgets/wxWidgets/blob/master/include/wx/defs.h#L2833

**Comment by vadz** [comment]
I forgot to say but, as usual, please let me know if you plan to make any changes or prefer to merge it like this (perhaps I'll add `MSWAdjustSizeOnDPIChange()` myself if I do it today) and leave any further improvements for later. TIA!

**Comment by MaartenBent** [comment]
I'll apply your suggestions and a better fix for the toolbar with `wxChoice`/`wxComboBox` control (it still randomly gets an incorrect height). I'll leave further improvements for later.

`wx/dataview.h` has some tabs replaced with spaces, you can revert this or split it into a separate commit. I won't do it, because I have to force push, and then reviewing my new changes becomes inconvenient.

**Comment by vadz** [comment]
Unfortunately there is a problem with missing `#include "wx/combobox.h"` with the latest version. Not sure if I should commit it or if you will.

**Comment by MaartenBent** [comment]
Oh, I didn't test the PCH-less build. Can you add it with merging? Thanks.

I had to replace `dynamic_cast<wxChoiceBase*>(control)` with `wxDynamicCast(control, wxComboBox) || wxDynamicCast(control, wxChoice)` because `wxDynamicCast` does not work exactly the same as `dynamic_cast`.

**Comment by vadz** [comment]
Yes, I've realized this -- we pretend that `wxComboBox` doesn't inherit from `wxChoice` under MSW because it doesn't do it in (all) the other ports.

I'll fix and merge it then.
